lib/tree.rb

```diff
@@ -2,61 +2,130 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
+
+  def add(key, val)
+    if key <= @key
+      if !left
+        @left = TreeNode.new(key, val)
+      else
+        @left.add(key, val)
+      end
+    else
+      if !right
+        @right = TreeNode.new(key, val)
+      else
+        @right.add(key, val)
+      end
+    end
+  end
+
+  def find(key)
+    if key == @key
+      return @value
+    elsif key < @key
+      return nil if !@left
+      return @left.find(key)
+    else
+      return nil if !@right
+      return @right.find(key)
+    end
+  end
+
+  def inorder(tree)
+    left.inorder(tree) unless !left
+    tree << {key: @key, value: @value}
+    right.inorder(tree) unless !right
+  end
+
+  def preorder(tree)
+    tree << {key: @key, value: @value}
+    left.preorder(tree) unless !left
+    right.preorder(tree) unless !right
+  end
+
+  def postorder(tree)
+    left.postorder(tree) unless !left
+    right.postorder(tree) unless !right
+    tree << {key: @key, value: @value}
+  end
+
+  def height(node)
+    return 0 if !node
+    left_height = height(node.left)
+    right_height = height(node.right)
+    max_height = left_height >= right_height ? left_height : right_height
+    return max_height + 1
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
```

Comment:
O(n) if the tree is unbalanced, O(log n) if it's balanced (average case approaches O(log n))